[dtensor] implement scaled dot product attention (flash-attention)

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120298
* #120297

as titled, this PR implements the sdpa flash attention op in DTensor

Adding flash attention first but efficient attention and other attention
ops should be similar

fixes https://github.com/pytorch/pytorch/issues/120333

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225

wanchaol committed:
[dtensor] implement scaled dot product attention (flash-attention)

as titled, this PR implements the sdpa flash attention op in DTensor

Adding flash attention first but efficient attention and other attention
ops should be similar

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] implement scaled dot product attention (flash-attention)"


as titled, this PR implements the sdpa flash attention op in DTensor

Adding flash attention first but efficient attention and other attention
ops should be similar

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

tianyu-l commented on test/distributed/_tensor/test_matrix_ops.py:
```diff
@@ -264,6 +265,46 @@ def test_placement_comb(
+        from torch.nn.attention import sdpa_kernel, SDPBackend
+
+        with sdpa_kernel(backends=[SDPBackend.FLASH_ATTENTION]):
+            out = F.scaled_dot_product_attention(
+                query, key, value, dropout_p=0.0, is_causal=True
+            )
+            dist_out = F.scaled_dot_product_attention(
+                dist_query, dist_key, dist_value, dropout_p=0.0, is_causal=True
+            )
```
do we want to check output are the same?

wanchaol replied:
added

tianyu-l commented on torch/distributed/_tensor/ops/matrix_ops.py:
```diff
@@ -128,3 +141,86 @@ def bmm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+        assert len(spec_list) == 6
+        input_expected_specs = spec_list[3:]
+        output_specs: List[Optional[DTensorSpec]] = list(spec_list[:3])
+        # fix up ouput_specs and fill in None for the int and empty tensor return values
+        for i in range(2, 8):
+            output_specs.insert(i, None)
```
Would you please share a pointer to the specification (input, output) of the flash attention op? I only found [this](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/FlashAttentionKernel.cpp#L146) which it doesn't seem to match.

wanchaol replied:
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L14526

tianyu-l reviewed (commented):
I wonder before this PR, what was the status of dtensor support for `scaled_dot_product_attention`? IIUC, in test_tp_examples.py we have a small transformer which already has F.sdpa computation compatible with dtensor.

wanchaol commented:
> before this PR, what was the status of dtensor support for scaled_dot_product_attention? IIUC, in test_tp_examples.py we have a small transformer which already has F.sdpa computation compatible with dtensor.

@tianyu-l We didn't support this op at all before. Our tp_examples using TP APIs, i.e. ColwiseParallel and RowwiseParallel, where in the attention layer after ColwiseParallel, it would automatically calls `to_local()` to convert to torch.Tensor, so scaled_dot_product_attention was not using DTensor

wanchaol committed:
Update on "[dtensor] implement scaled dot product attention (flash-attention)"


as titled, this PR implements the sdpa flash attention op in DTensor

Adding flash attention first but efficient attention and other attention
ops should be similar

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

XilunWu reviewed (approved):
lgtm. Left a question about RNG.

wanchaol committed:
Update on "[dtensor] implement scaled dot product attention (flash-attention)"


as titled, this PR implements the sdpa flash attention op in DTensor

Adding flash attention first but efficient attention and other attention
ops should be similar

fixes https://github.com/pytorch/pytorch/issues/120333

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol commented on test/distributed/_tensor/test_matrix_ops.py:
```diff
@@ -264,6 +265,61 @@ def test_placement_comb(
+            )
+            full_out = dist_out.full_tensor()
+            if self.rank == 0:
+                self.assertEqual(full_out, out)
```
It was mainly my debugging and forgot to toggle it, we can do a non-rank sensitive assertequal for this case, updated

wanchaol commented:
@pytorchbot merge
